graph/preprocessor.go

```diff
@@ -0,0 +1,320 @@
+			ongoingCmd = true
+		} else {
+			out.WriteRune(char)
+		}
+	}
+
+	return out.String(), changed, nil
+}
+
+// preprocessBytes handles byte encoded data that can be parsed through pre processing
+func preprocessBytes(data []byte) ([]byte, Alias, bool, error) {
+	type Wrapper struct {
+		Alias Alias `yaml:"alias,omitempty"`
+	}
+
+	wrap := &Wrapper{}
+	aliasData, remainingData := basicAliasSeparation(data)
+
+	if errUnmarshal := yaml.Unmarshal(aliasData, wrap); errUnmarshal != nil {
+		return data, Alias{}, false, errUnmarshal
+	}
+
+	alias := &wrap.Alias
+
+	if alias.AliasMap == nil {
+		// Alias Src defined. Guarantees alias map can be populated
+		alias.AliasMap = make(map[string]string)
+	}
+	// Search and Replace
+	str := string(remainingData)
+	parsedStr, changed, err := preprocessString(alias, str)
+
+	return []byte(parsedStr), *alias, changed, err
+}
+
+// processSteps Will resolve image names in steps that are aliased without using directive.
+// Invoked after resolving all directive using aliases
+func processSteps(alias *Alias, task *Task) {
+	for i, step := range task.Steps {
+		parts := strings.Split(strings.TrimSpace(step.Cmd), " ")
+		if val, ok := alias.AliasMap[parts[0]]; ok {
+			// Image name should always go first
+			parts[0] = val
+			task.Steps[i].Cmd = strings.Join(parts, " ")
+		}
+	}
+}
+
+// Provides simple separation of the top level items in a yaml file definition, however, need
+// to update to be fully compliant (include JSON top level for example). Alternative construction
+// of small compiler for this purpose is also under consideration.
+func basicAliasSeparation(data []byte) ([]byte, []byte) {
+	reader := bytes.NewReader(data)
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(bufio.ScanLines)
+
+	var aliasBuffer bytes.Buffer
+	var buffer bytes.Buffer
+
+	inside := false
+	aliasFieldName := regexp.MustCompile(`\Aalias\s*:.*\z`)
+	genericTopLevelRe := regexp.MustCompile(`\A[^\s:]+[^:]*:.*\z`)
+	commentRe := regexp.MustCompile(`\A\s*#.*`)
+	for scanner.Scan() {
+		text := scanner.Text()
+		if matched := commentRe.MatchString(text); matched {
+			continue
+		}
+
+		if matched := aliasFieldName.MatchString(text); matched && !inside {
+			inside = true
+		} else if matched := genericTopLevelRe.MatchString(text); matched && inside {
+			inside = false
+		}
+
+		if inside {
+			aliasBuffer.WriteString(text + "\n")
+		} else {
+			buffer.WriteString(text + "\n")
+		}
+	}
+
+	return aliasBuffer.Bytes(), buffer.Bytes()
+}
+
+// FindVersion Determines the current version of an Alias task file
+func FindVersion(data []byte) string {
+	reader := bytes.NewReader(data)
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(bufio.ScanLines)
+	commentRe := regexp.MustCompile(`\A\s*#.*`)
+	versionField := regexp.MustCompile(`\Aversion\s*:.+\z`)
+
+	for scanner.Scan() {
+		text := scanner.Text()
+
+		//Finds the first non empty and non comment string
+		comment := commentRe.MatchString(text)
```

Comment:
Have you considered whitespace? In your case a line containing whitespace will `break`. 

Remove usage of comments and version regex. Do this instead:

```go
for scanner.Scan()  {
  text := scanner.Text()

  trimmedText = strings.TrimSpace(text)

  // Skip comments and just whitespace.
  if trimmedText == "" || strings.HasPrefix(trimmedText, "#") {
    continue
  }

  // use text instead of trimmedText since '   version: ' is also invalid. 
  if strings.HasPrefix(text, "version:") {
    return strings.TrimPrefix(trimmedText, "version:")
  }
  break
}
```

Reply:
Nice solution, I like it